.circleci/config.yml

```diff
@@ -1,6 +1,19 @@
 version: 2.1
 
 jobs:
+  controller-test:
+    machine:
+      image: ubuntu-1604:201903-01
+    steps:
+      - checkout
+      - run:
+          name: Run Docker build
+          command: |
+            docker build -f controller/Dockerfile -t controller .
```

Comment:
Does this happen every time tests are run or is it somehow cached? If it happens every time, what do we get out of using Docker here - I thought the point of Docker in such a situation would be to have a prebuilt test environment?

Reply:
The main thing we get out of Docker is a guarantee that if it builds in Docker, it works everywhere, and gives us a consistent environment to eventually build releases in.

In the future, when there's not so much churn in the dependencies, we can make builder layer images to speed this up. For the time being though, a few minutes building a container feels like a good tradeoff for having a release environment out of the box.

In addition, we're using docker_layer_caching in circle, so we'll be able to reuse images across jobs, at least.

Reply:
I imagine it would be fairly easy to cache the image or use a multistage build. I can add that.